cli/lsp/diagnostics.rs

```diff
@@ -25,8 +37,317 @@ pub enum DiagnosticSource {
+        disturbed = true;
+      }
+    }
+    Err(err) => {
+      error!("Internal error: {}", err);
+    }
+    _ => (),
+  }
+  performance.measure(mark);
+
+  if disturbed {
+    publish_diagnostics(client, collection, snapshot, performance).await
+  }
+}
+
+fn handle_request(
+  maybe_request: Option<DiagnosticRequest>,
+  collection: &mut DiagnosticCollection,
+  dirty: &mut bool,
+) -> bool {
+  match maybe_request {
+    Some(request) => {
+      match request {
+        DiagnosticRequest::Get(specifier, source, tx) => {
+          let diagnostics = collection
+            .diagnostics_for(&specifier, &source)
+            .cloned()
+            .collect();
+          if tx.send(diagnostics).is_err() {
+            error!("DiagnosticServer unable to send response on channel.");
+          }
+        }
+        DiagnosticRequest::Invalidate(specifier) => {
+          collection.invalidate(&specifier)
+        }
+        DiagnosticRequest::Update => *dirty = true,
+      }
+      true
+    }
+    _ => false,
+  }
+}
+
+/// A server which calculates diagnostics in its own thread and publishes them
+/// to an LSP client.
+#[derive(Debug)]
+pub(crate) struct DiagnosticsServer(
+  Option<mpsc::UnboundedSender<DiagnosticRequest>>,
+);
+
+impl DiagnosticsServer {
+  pub(crate) fn new() -> Self {
+    Self(None)
+  }
+
+  pub(crate) fn start(
+    &mut self,
+    language_server: Arc<tokio::sync::Mutex<language_server::Inner>>,
+    client: Client,
+    ts_server: Arc<tsc::TsServer>,
+    performance: Performance,
+  ) {
+    let (tx, mut rx) = mpsc::unbounded_channel::<DiagnosticRequest>();
+    self.0 = Some(tx);
+
+    let _join_handle = thread::spawn(move || {
+      let runtime = create_basic_runtime();
+      let mut collection = DiagnosticCollection::default();
+
+      runtime.block_on(async {
+        // Some(snapshot) is the flag we use to determine if something has
+        // changed where we will wait for the timeout of changes or a request
+        // that forces us to update diagnostics
+        let mut dirty = false;
+
+        loop {
+          let next = rx.recv();
+          tokio::pin!(next);
+
+          let duration = if dirty {
+            time::Duration::from_millis(DIAGNOSTIC_DEBOUNCE_MS)
+          } else {
+            // we need to await an arbitrary silly amount of time, so this is
+            // 1 year in seconds
+            time::Duration::new(31_622_400, 0)
+          };
+          let debounce = time::sleep(duration);
+          tokio::pin!(debounce);
+
+          // "race" the next message off the rx queue or the debounce timer.
+          // if the next message comes off the queue, the next iteration of the
+          // loop will reset the debounce future.  When the debounce future
+          // occurs, the diagnostics will be updated based on only the most
+          // recent snapshot, thereby "abandoning" any interim states.
+          tokio::select! {
+            _ = &mut debounce => {
```

Comment:
I might be mistaken but I think you can skip the pinning and just do:
```suggestion
            _ = time::sleep(duration) => {
```